### PR TITLE
WIP: Divide train/test folder, validate on dev set instead of test set

### PR DESCRIPTION
**Pull request by titipata**
- The right way should be to divide training set into training set and development set. Later we can validate on the test set.
- Saving train and test to separate folder
- Adding `callback_lists` while training to save best model on development set

### CODE REVIEW

**Comment by rkcosmos** [comment]
The way you separate train/dev set 

x_train_char, x_dev_char, x_train_type, x_dev_type, y_train, y_dev = train_test_split(x_train_char, x_train_type, y_train, test_size=0.1)

will create information leakage where model see some information in dev set from training (for example, data from x_train_char and x_dev_char can be adjacent character from the same sentence. In this case, the model get some idea about x_dev_char while we train it. This will unfairly increase performance score in dev set)

The way to prevent information leakage is to create 3 dataset, train, validation and test set instead of 2 dataset in this step

files_train, files_test = train_test_split(files, random_state=0, test_size=0.1)   

then train on training set, validate on val set and report score from test set

**Comment by titipata** [comment]
@rkcosmos, ah, you're right. I'll fix that by creating 3 dataset instead.

**Comment by titipata** [comment]
I fixed on your comment  by have an option to divide `create_dev` in `generate_best_dataset`.  If there is a `dev` folder exist, we'll load that and use it as validation set. We'll again save the best weight on dev set. Let me know if this is right.

**Comment by rkcosmos** [comment]
Perfect!, there is one thing I want to discuss about convention. n_gram is a standard word in nlp, so I think we should call our variable n_gram rather than n_pad. Also I would normally call val_set instead of dev_set. 

My experience in the field is actually limited so I'm not sure if n_pad or dev_set is the convention as well or not.

**Comment by titipata** [comment]
`n_gram` is for more like when you have `n_gram` of words e.g. unigram, bigram, or trigram of words. I haven't seen people use it for character. 

Yeah, `dev_set` and `val_set` is depended on the field. For Data Science, people use `val_set` and `test_set`. For computer science people, they use `dev_set` to describe the same thing, haha.

I'd say I'll just use `n_pad` for padding characters cause we're not padding the words. However, `dev_set` and `val_set` are interchangeable. I can change it to `val_set` too. I kind of like it better and convey information better than development set.

**Comment by titipata** [comment]
@rkcosmos, I change to `val_set` cause I think it makes more sense in this case :) I still think that `n_pad` is a better name for this task more than `n_gram` tho so I didn't change it.
